### PR TITLE
Bug 1807204: Update cursor to show drop status over targets

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
**Fixes**: 
https://issues.redhat.com/browse/ODC-3028

**Analysis / Root cause**: 
The cursor was not updated during drag operations to indicate drop status.

**Solution Description**: 
Modified the cursor to:
- Use an SVG for the Add cursor when creating connections
- Show a tooltip when over a valid add drop target informing the user that dropping will add an application here (after the standard delay without a mouse move)
- Show an invalid drop cursor when over items where a drop would do nothing

** Sample Screen Shots **
Drag_Cursor_2

**Browser conformance**: 
- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

cc @openshift/team-devconsole-ux 

/kind bug

### CODE REVIEW

File: frontend/packages/topology/src/behavior/withCreateConnector.tsx [hunk 1]
```diff
@@ -135,6 +135,22 @@ const CreateConnectorWidget: React.FC<CreateConnectorWidgetProps> = observer((pr
...
+  const unsetHandle = React.useRef<number>();
+
+  React.useEffect(() => {
+    setHover(false);
+    clearTimeout(unsetHandle.current);
+    unsetHandle.current = window.setTimeout(() => setHover(true), 2000);
+    return () => {
+      clearTimeout(unsetHandle.current);
+    };
+  }, [endPoint.x, endPoint.y]);
```
Comment (christianvogt) [inline frontend/packages/topology/src/behavior/withCreateConnector.tsx]: This is better off going directly into the `DefaultCreateConnector`.

In dev-console we can supply our own prop to `renderConnector` which uses `DefaultCreateConnector` to supply the custom message.

File: frontend/packages/dev-console/src/components/topology/components/nodes/KnativeServiceGroup.tsx [hunk 2]
```diff
@@ -92,7 +92,12 @@ const KnativeServiceGroup: React.FC<KnativeServiceGroupProps> = ({
   }, [editAccess, innerHover, onShowCreateConnector, onHideCreateConnector]);
 
   return (
-    <Tooltip content="Move sink to service" trigger="manual" isVisible={dropTarget && canDrop}>
+    <Tooltip
+      content="Move sink to service"
+      trigger="manual"
+      isVisible={dropTarget && canDrop}
+      tippyProps={{ duration: 0, delay: 0 }}
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/components/nodes/KnativeServiceGroup.tsx]: Can you confirm the behavior with UX? 0 delay will bring up tooltips immediately as you hover your mouse over an intermediate node.

Didn't we try avoid this kind of behavior for label truncation by having a small delay.

File: frontend/packages/dev-console/src/components/topology/componentUtils.ts [hunk 3]
```diff
@@ -174,9 +174,20 @@ const graphWorkloadDropTargetSpec: DropTargetSpec<
         monitor.getItemType() === CREATE_CONNECTOR_DROP_TYPE)
     );
   },
-  collect: (monitor) => ({
-    dragEditInProgress: monitor.isDragging() && editOperations.includes(monitor.getOperation()),
-  }),
+  collect: (monitor) => {
+    const dragEditInProgress =
+      monitor.isDragging() && editOperations.includes(monitor.getOperation());
+    const dropHints = monitor.getDropHints() || [];
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/topology/componentUtils.ts]: maybe we should change `getDropHints` to always return an array (`string[]`)?

File: frontend/packages/dev-console/src/components/topology/components/nodes/KnativeServiceGroup.tsx (hunk 2 above)
Comment (jeff-phillips-18) [reply frontend/packages/dev-console/src/components/topology/components/nodes/KnativeServiceGroup.tsx]: UX spec specifies `For connectors, the tooltip should appear the moment the user hovers over a valid drop target.`

**Comment by jeff-phillips-18** [comment]
Updated per comments.

Removed the `grab` cursor when dragging edges over valid drop nodes.

In order to move out the dev-console specific hover treatment for the create connector, I had to refactor a bunch to make the default connector a component rather than a function call that returns a component.

File: frontend/packages/topology/src/components/DefaultCreateConnector.tsx [hunk 4]
```diff
@@ -1,40 +1,66 @@
 import * as React from 'react';
-import Point from '../geom/Point';
+import * as classNames from 'classnames';
+import { AddCircleOIcon } from '@patternfly/react-icons';
+import { Tooltip } from '@patternfly/react-core';
 import ConnectorArrow from './ConnectorArrow';
+import { CreateConnectorRenderer } from './CreateConnectorWidget';
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/DefaultCreateConnector.tsx]: `CreateConnectorWidget` imports `DefaultCreateConnector`
and `DefaultCreateConnector` imports  `CreateConnectorWidget`
Probably worth combining the two together.

File: frontend/packages/topology/src/components/CreateConnectorWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+import { AnchorEnd, Graph, GraphElement, isGraph, isNode, Node } from '../types';
+import Point from '../geom/Point';
+import Layer from './layers/Layer';
+import { hullPath } from '../utils';
+import { ContextMenu, ContextMenuItem } from './contextmenu';
+
+import './CreateConnectorWidget.scss';
+
+export const CREATE_CONNECTOR_OPERATION = 'createconnector';
+export const CREATE_CONNECTOR_DROP_TYPE = '#createConnector#';
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/CreateConnectorWidget.tsx]: Old code but:

```suggestion
export const CREATE_CONNECTOR_OPERATION = '#createconnector#';
export const CREATE_CONNECTOR_DROP_TYPE = '#createConnector#';
```

File: frontend/packages/topology/src/components/CreateConnectorWidget.tsx [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+export type CreateConnectorRendererProps = {
+  startPoint: Point;
+  endPoint: Point;
+  dragging: boolean;
+  hints: string[] | undefined;
+  tipContents?: React.ReactNode;
+  className?: string;
+};
+
+export type CreateConnectorRenderer = React.ComponentType<CreateConnectorRendererProps>;
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/CreateConnectorWidget.tsx]: This isn't correct. The only required props that are expected of a component used in `withCreateConnector` are:
```
  startPoint: Point,
  endPoint: Point,
  hints: string[] | undefined,
```

File: frontend/packages/topology/src/components/DefaultCreateConnector.tsx [hunk 7]
```diff
@@ -1,40 +1,66 @@
...
 
-type DefaultCreateConnectorProps = {
-  startPoint: Point;
-  endPoint: Point;
-  hints?: string[];
-};
+const cursorSize = 20;
 
-const DefaultCreateConnector: React.FC<DefaultCreateConnectorProps> = ({
+const DefaultCreateConnector: CreateConnectorRenderer = ({
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/DefaultCreateConnector.tsx]: You'll need a new prop type that includes support for `tipContents` and `className`.

File: frontend/packages/topology/src/components/index.ts [hunk 8]
```diff
@@ -5,3 +5,5 @@ export { default as ConnectorArrow } from './ConnectorArrow';
 export { default as EdgeConnectorArrow } from './EdgeConnectorArrow';
 export { default as GraphComponent } from './GraphComponent';
 export { default as VisualizationSurface } from './VisualizationSurface';
+export * from './CreateConnectorWidget';
```
Comment (christianvogt) [inline frontend/packages/topology/src/components/index.ts]: Do not expose contents of `CreateConnectorWidget` as API. This is for internal use only to `withCreateConnector`. If needed, re-export any specific types from `withCreateConnector`.

**Pull request merged**